src/rest/connector/libs/bigip/implementation.py

```diff
@@ -163,107 +156,148 @@ def connect(self, auth_provider='tmos', verify=False, port='443', protocol='http
+                                                          ip=self.ip,
+                                                          port=self.port)
 
         self.username, self.password = get_username_password(self)
 
         self.header = "Content-Type: application/json"
         self.verify = verify
+        self._auth_provider = auth_provider
+        self._ttl = ttl
 
-        # URL to authenticate and receive the token
-        url = "https://{0}:{1}/mgmt/shared/authn/login".format(
-            ip, port
+        self._connect(timeout, retries, retry_wait)
+
+        return self._is_connected, self.icr_session
+
+    def _connect(self, timeout: int, retries: int, retry_wait: int):
+        """ Authenticate and initiate a session with the device
+
+        Args:
+            timeout: The timeout to use when establishing the connection
+            retries: How many times to retry to connect to the device if it fails
+            retry_wait: Time in seconds to wait between retries
+        """
+        self._authenticate(timeout, retries, retry_wait)
+
+        self._extend_session_ttl(self._ttl)
+
+        params = dict(
+            username=self.username,
+            password=self.password,
+            verify=self.verify,
+            token_to_use=self.token
+        )
+
+        # creating an object to be used all new requests
+        self.icr_session = iControlRESTSession(**params)
+
+        self._is_connected = True
+
+        log.info(
+            "Connected successfully to '%s'", self.device.name
         )
+
+    def _authenticate(self, timeout: int, retries: int, retry_wait: int):
+        """ Authenticates with the device and retrieves a session token to be
+            used in actual requests
+
+        Args:
+            timeout: The timeout to use when establishing the connection
+            retries: How many times to retry to connect to the device if it fails
+            retry_wait: Time in seconds to wait between retries
+        """
+        # URL to authenticate and receive the token
+        url = f"{self.base_url}/mgmt/shared/authn/login"
+
         payload = {
             'username': self.username,
             'password': self.password,
-            'loginProviderName': auth_provider
+            'loginProviderName': self._auth_provider
         }
 
         iCRS = iControlRESTSession(
             self.username,
             self.password,
+            timeout=timeout,
             verify=self.verify
         )
 
         log.info(
-            "Connecting to '{d}' with alias "
-            "'{a}'".format(d=self.device.name, a=self.alias)
+            "Connecting to '%s'", self.device.name
         )
 
         response = iCRS.post(
             url,
             json=payload,
         )
 
-        log.info(response.json())
+        log.debug(response.json())
 
-        if response.status_code not in [200]:
+        if response.status_code != 200:
```

Comment:
I'm thinking it's better to leave this as-is. Functionally they're the same, but the former leaves room for addition in the future if necessary (and indicates to future contributors that it is acceptable to add to it)

Reply:
This is a really suboptimal implementation though, as using lists inline to test for membership imply the JIT allocation of the list, only to traverse it and then destroy it immediately.

In this case it's particularly outstanding as we really only want to check for HTTP 200 (OK), any other HTTP code will effectively mean that the operation failed.

If we really want to keep that though, I would suggest defining a tuple at the module scope, and then using that for membership testing throughout the code instead, as it's 1) more memory efficient and 2) faster as you don't have to allocate and deallocate every time 3) less error prone, no need to redefine the container over and over again